x-pack/plugin/identity-provider/src/test/java/org/elasticsearch/xpack/idp/action/SamlIdentityProviderTests.java

```diff
@@ -0,0 +1,350 @@
+        final String nameIdFormat = TRANSIENT;
+        final String relayString = null;
+        final boolean forceAuthn = randomBoolean();
+        final AuthnRequest authnRequest = buildAuthnRequest(entityId + randomAlphaOfLength(4), new URL(acsUrl),
+            new URL("https://idp.org/sso/redirect"), nameIdFormat, forceAuthn);
+        final String query = getQueryString(authnRequest, relayString, false, null);
+        validateRequest.setJsonEntity("{\"authn_request_query\":\"" + query + "\"}");
+        ResponseException e = expectThrows(ResponseException.class, () -> getRestClient().performRequest(validateRequest));
+        assertThat(e.getMessage(), containsString("is not registered with this Identity Provider"));
+        assertThat(e.getResponse().getStatusLine().getStatusCode(), equalTo(RestStatus.BAD_REQUEST.getStatus()));
+    }
+
+    public void testSpInitiatedSsoFailsForMalformedRequest() throws Exception {
+        String acsUrl = "https://" + randomAlphaOfLength(12) + ".elastic-cloud.com/saml/acs";
+        String entityId = SP_ENTITY_ID;
+        assertServiceProviderIsRegistered(entityId, acsUrl);
+        ensureGreen(SamlServiceProviderIndex.INDEX_NAME);
+        // Validate incoming authentication request
+        Request validateRequest = new Request("POST", "/_idp/saml/validate");
+        validateRequest.setOptions(IDP_REQUEST_OPTIONS);
+        final String nameIdFormat = TRANSIENT;
+        final String relayString = null;
+        final boolean forceAuthn = randomBoolean();
+        final AuthnRequest authnRequest = buildAuthnRequest(entityId + randomAlphaOfLength(4), new URL(acsUrl),
+            new URL("https://idp.org/sso/redirect"), nameIdFormat, forceAuthn);
+        final String query = getQueryString(authnRequest, relayString, false, null);
+        // Skip http parameter name
+        final String queryWithoutParam = query.substring(12);
+        validateRequest.setJsonEntity("{\"authn_request_query\":\"" + queryWithoutParam + "\"}");
+        ResponseException e = expectThrows(ResponseException.class, () -> getRestClient().performRequest(validateRequest));
+        assertThat(e.getMessage(), containsString("does not contain a SAMLRequest parameter"));
+        assertThat(e.getResponse().getStatusLine().getStatusCode(), equalTo(RestStatus.BAD_REQUEST.getStatus()));
+        // arbitrarily trim the request
+        final String malformedRequestQuery = query.substring(0, query.length() - randomIntBetween(10, 15));
+        validateRequest.setJsonEntity("{\"authn_request_query\":\"" + malformedRequestQuery + "\"}");
+        ResponseException e1 = expectThrows(ResponseException.class, () -> getRestClient().performRequest(validateRequest));
+        assertThat(e1.getResponse().getStatusLine().getStatusCode(), equalTo(RestStatus.BAD_REQUEST.getStatus()));
+    }
+
+    private void assertServiceProviderIsRegistered(String entityId, String acsUrl) throws Exception {
```

Comment:
This isn't really an `assert` - it's more of 
```suggestion
    private void registerServiceProvider(String entityId, String acsUrl) throws Exception {
```

Reply:
Well it does assert that the service provider is registered after it registers it, but I see your point

Reply:
Ahem! I'm supposed to be the pedant on the team. I take a couple of days off, and what happens?

Reply:
pedant-stepping in, like a boss :D